File: C++/tests/group_anagrams_test.cpp
#include <algorithm>
#include <group_anagrams.hpp>
#include <gtest/gtest.h>
#include <string>
#include <vector>

using namespace std;

class GroupAnagramsTest : public ::testing::Test {
protected:
  Solution solution;

  void check_equal(const vector<vector<string>> &result,
                   const vector<vector<string>> &want) {
    auto sort_groups = [](auto &vec) {
      for (auto &group : vec) {
        sort(group.begin(), group.end());
      }
      sort(vec.begin(), vec.end());
    };

    auto result_copy = result;
    auto want_copy = want;

    sort_groups(result_copy);
    sort_groups(want_copy);

    ASSERT_EQ(result_copy, want_copy);
  }
};

TEST_F(GroupAnagramsTest, same_length) {
  vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
  vector<vector<string>> want = {
      {"bat"}, {"nat", "tan"}, {"ate", "eat", "tea"}};
  check_equal(solution.groupAnagrams(strs), want);
}

TEST_F(GroupAnagramsTest, empty_string) {
  check_equal(solution.groupAnagrams({""}), {{""}});
}

TEST_F(GroupAnagramsTest, single_character_strings) {
  check_equal(solution.groupAnagrams({"a", "b", "a"}), {{"a", "a"}, {"b"}});
}

TEST_F(GroupAnagramsTest, mixed_case_strings) {
  check_equal(solution.groupAnagrams({"a", "A"}), {{"a"}, {"A"}});
}

TEST_F(GroupAnagramsTest, no_anagrams) {
  check_equal(solution.groupAnagrams({"abc", "def", "ghi"}),
              {{"abc"}, {"def"}, {"ghi"}});
}

TEST_F(GroupAnagramsTest, all_anagrams) {
  check_equal(solution.groupAnagrams({"abc", "bca", "cab"}),
              {{"abc", "bca", "cab"}});
}

File: JavaScript/tests/group_anagrams.test.js
import { groupAnagrams } from "#src/group_anagrams.js";
import { expect, test } from "vitest";

const cases = [
  {
    name: "same length",
    strs: ["eat", "tea", "tan", "ate", "nat", "bat"],
    want: [["bat"], ["nat", "tan"], ["ate", "eat", "tea"]],
  },
  {
    name: "empty string",
    strs: [""],
    want: [[""]],
  },
  {
    name: "single character strings",
    strs: ["a", "b", "a"],
    want: [["a", "a"], ["b"]],
  },
  {
    name: "mixed case strings",
    strs: ["a", "A"],
    want: [["a"], ["A"]],
  },
  {
    name: "no anagrams",
    strs: ["abc", "def", "ghi"],
    want: [["abc"], ["def"], ["ghi"]],
  },
  {
    name: "all anagrams",
    strs: ["abc", "bca", "cab"],
    want: [["abc", "bca", "cab"]],
  },
];

for (const tt of cases) {
  test(tt.name, () => {
    const actual = groupAnagrams(tt.strs);
    actual.forEach((arr) => arr.sort());
    tt.want.forEach((arr) => arr.sort());
    actual.sort((a, b) => a.length - b.length);
    tt.want.sort((a, b) => a.length - b.length);
    expect(actual).toEqual(tt.want);
  });
}
